modules/engine-prototype/src/main/scala/engine/Sequence.scala

```diff
@@ -0,0 +1,91 @@
+package engine
+
+import scala.concurrent.ExecutionContext
+
+import cats.{FlatMap, Functor}
+import cats.data.NonEmptyList
+import cats.effect.Effect
+import cats.effect.implicits._
+import cats.implicits._
+
+import fs2.Stream
+import fs2.async
+import fs2.async.mutable.Signal
+
+import monocle.{Lens, Prism}
+import monocle.macros.GenLens
+
+sealed trait Sequence
+object Sequence {
+
+  sealed trait F2 extends Sequence
```

Comment:
A Sequence is an ADT supporting different instruments: `Sequence = Sequence.F2 | Sequence.GMOS | ...`. This way we ensure we can create Sequences that have Steps that are supported by the Sequence. For example, you shouldn't be able to create a Sequence with configurations for GMOS and F2.

Reply:
I agree the way ADTs are written in the code is cumbersome but are subtypes the only option to make this kind of ADTs easier to write. Aren't there more options?

Writing this kind of ADTs in Haskell is trivial so I'm pretty sure there must be some other reasonable ways to encode nested ADTs easily in Scala. I'm missing reading more examples of encoding Scala ADTs in the wild, what we have here is not that special.